### ISSUE TITLE
Ideas for client interface

### ISSUE THREAD

**Issue by uNetworkingAB** [uNetworking/uWebSockets#1527]
Been thinking about how absurd it is to shoehorn a client into the URL router. Better interface would be something like:

#include "Client.h"

```
uWS::Client({
.open = [](auto *ws) {

},
.message = [](auto *ws, auto message) {

}
.close = [](auto *ws) {

}
}).connect("wss://someplace.com");
```

Much simpler and totally isolated from the rest.

**Comment by cirospaciari** [comment]
I think adding protocols as  optional on connect to keep client compatibility with browser will be great https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_client_applications

open, messages, close and maybe add error to deal with connection errors or just pass it to close handler 

An simple an isolated client approach make completely sense.

**Comment by uNetworkingAB** [comment]
Yep something like error makes sense for client but not for server

**Comment by woater** [comment]
Does interfaces of WebSockets of this lib only support construct servers but not clients now?

**Comment by HagaiVinik** [comment]
Hello, joining the discussion, about the new `ClientApp`:
Is it a ready for production code? I saw it's not in the release yet but still though it is worth to ask (because it will be very useful for me) @uNetworkingAB

**Comment by cirospaciari** [comment]
@HagaiVinik is just an Idea, for now, you can see the code is just a mock https://github.com/uNetworking/uWebSockets/blob/master/src/ClientApp.h
@woater WebSockets is focused on servers, not WS clients yet, but the idea is to implement a client, this is the idea for the interface.

**Comment by HagaiVinik** [comment]
@cirospaciari 
Appreciate the quick answer mate, currently I need to proxy a request i have received on my uWS server to a different host, and proxy his response back. I tried to look through the framework and try using the `ws` method. (webSocket module?) 
What is your suggestion for implementing this the best way? is there a similar example I can take inspiration from?

**Comment by cirospaciari** [comment]
> @cirospaciari Appreciate the quick answer mate, currently I need to proxy a request i have received on my uWS server to a different host, and proxy his response back. I tried to look through the framework and try using the `ws` method. (webSocket module?) What is your suggestion for implementing this the best way? is there a similar example I can take inspiration from?

Personally, I have a project that can wait for uWS, but if is not your case this are my suggestions:
 
I needed to do something similar to scaling, I used a message broker in this case, Redis pub/sub, another alternative will be RabbitMQ, but Redis pub/sub can do the job very well. 

In my case, I need this client because I want to use UNIX domain socket to communicate internally, without any other library. If you cannot use a message broker, I will suggest you take a look in some other native libraries like libwebsockets, and do some benchmarks, load tests to see if you can rely on it for client, until uWS get the client working.

**Comment by HagaiVinik** [comment]
> 

@cirospaciari 
Thanks for answering.
to be honest those solutions are kind of `overpowered` for me.
I am trying to use the library somehow make it generate a connection with other server.
Do you have any suggestions regarding this mechanism?

**Comment by cirospaciari** [comment]
> > 
> 
> @cirospaciari Thanks for answering. to be honest those solutions are kind of 'overpowered' for me. I am trying to use the uWS library somehow and make it generate a connection with a other server. Do you have any suggestions regarding this mechanism?

@HagaiVinik @woater  I opened this discussion let's go to there https://github.com/uNetworking/uWebSockets/discussions/1534

**Comment by jwang0512** [comment]
It's great to hear uWS will have client api. can not waiting for it 👍

**Comment by uNetworkingAB** [comment]
That's speculation - this is just a simple interface idea. It literally says nothing about how, if, when, at what price, with what license, this might be implemented, if ever.
